tests/branch_management/tests/test_branches.py

```diff
@@ -1,93 +1,127 @@
 #
 # Copyright 2024 Canonical, Ltd.
 #
+import functools
+import logging
+import subprocess
 from pathlib import Path
-from subprocess import check_output
 
 import requests
 
+log = logging.getLogger(__name__)
+K8S_GH_REPO = "https://github.com/canonical/k8s-snap.git/"
+K8S_LP_REPO = " https://git.launchpad.net/k8s"
 
-def _get_max_minor(major):
-    """Get the latest minor release of the provided major.
-    For example if you use 1 as major you will get back X where X gives you latest 1.X release.
-    """
-    minor = 0
-    while _upstream_release_exists(major, minor):
-        minor += 1
-    return minor - 1
-
-
-def _upstream_release_exists(major, minor):
-    """Return true if the major.minor release exists"""
-    release_url = "https://dl.k8s.io/release/stable-{}.{}.txt".format(major, minor)
-    r = requests.get(release_url)
-    return r.status_code == 200
 
-
-def _confirm_branch_exists(branch):
-    cmd = f"git ls-remote --heads https://github.com/canonical/k8s-snap.git/ {branch}"
-    output = check_output(cmd.split()).decode("utf-8")
-    assert branch in output, f"Branch {branch} does not exist"
+def _sh(*args, **kwargs):
+    default = {"text": True, "stderr": subprocess.PIPE}
+    try:
+        return subprocess.check_output(*args, **{**default, **kwargs})
+    except subprocess.CalledProcessError as e:
+        log.error("stdout: %s", e.stdout)
+        log.error("stderr: %s", e.stderr)
+        raise e
 
 
 def _branch_flavours(branch: str = None):
     patch_dir = Path("build-scripts/patches")
     branch = "HEAD" if branch is None else branch
-    cmd = f"git ls-tree --full-tree -r --name-only {branch} {patch_dir}"
-    output = check_output(cmd.split()).decode("utf-8")
+    cmd = f"git ls-tree --full-tree -r --name-only origin/{branch} {patch_dir}"
+    output = _sh(cmd.split())
     patches = set(
         Path(f).relative_to(patch_dir).parents[0] for f in output.splitlines()
     )
     return [p.name for p in patches]
 
 
-def _confirm_recipe(track, flavour):
+@functools.lru_cache
+def _confirm_branch_exists(repo, branch):
+    log.info(f"Checking {branch} branch exists in {repo}")
+    cmd = f"git ls-remote --heads {repo} {branch}"
+    output = _sh(cmd.split())
+    return branch in output
```

Comment:
the branches shouldn't be changing while we're checking them.  lets cache the results